> @prajwalthakur rebase on main to get CI to turn over (or if it still fails, likely due to an issue your work has still) 

I rebased the main (7e382cb) commit, and add my changes on it (two files in nav2_mppi_controller), ci is still failing, hmm not sure why..
its showing error while building the nav2_util 
```cc1plus: note: set ‘-fconcepts-diagnostics-depth=’ to at least 2 for more detail
/usr/include/c++/15/bits/stl_construct.h:133:7: error: no matching function for call to ‘tf2_ros::CreateTimerROS::CreateTimerROS(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, std::shared_ptr<rclcpp::node_interfaces::NodeTimersInterface>)’
  133 |       ::new(static_cast<void*>(__p)) _Tp(std::forward<_Args>(__args)...);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/15/bits/stl_construct.h:133:7: note: there is 1 candidate
In file included from /opt/overlay_ws/src/navigation2/nav2_util/src/base_footprint_publisher.hpp:24,
                 from /opt/overlay_ws/src/navigation2/nav2_util/src/base_footprint_publisher.cpp:17:
/opt/ros/rolling/include/tf2_ros/tf2_ros/create_timer_ros.hpp:64:3: note: candidate 1: ‘tf2_ros::CreateTimerROS::CreateTimerROS(RequiredInterfaces, rclcpp::CallbackGroup::SharedPtr)’
   64 |   CreateTimerROS(
      |   ^~~~~~~~~~~~~~
/opt/ros/rolling/include/tf2_ros/tf2_ros/create_timer_ros.hpp:66:38: note: no known conversion for argument 2 from ‘shared_ptr<rclcpp::node_interfaces::NodeTimersInterface>’ to ‘shared_ptr<rclcpp::CallbackGroup>’
   66 |     rclcpp::CallbackGroup::SharedPtr callback_group = nullptr);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [src/CMakeFiles/base_footprint_publisher.dir/build.make:79: src/CMakeFiles/base_footprint_publisher.dir/base_footprint_publisher.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:326: src/CMakeFiles/base_footprint_publisher.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< nav2_util [8.23s, exited with code 2]```

--- a/nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp
+++ b/nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp
@@ -167,6 +167,31 @@ public:
   }
 
   /**
+   * @brief With input pose, speed find the vehicle's output pose at pred_dt in future
+   * @param pose pose of the robot
+   * @param speed speed 
+   * @param pred_dt time in future to predict the pose
+  */
+  virtual void predictFuture(geometry_msgs::msg::PoseStamped& pose, geometry_msgs::msg::Twist& speed, float pred_dt)
+  {
+    const bool is_holo = isHolonomic();
+    auto initial_yaw = static_cast<float>(tf2::getYaw(pose.pose.orientation));
+    auto yaw_cos = cosf(initial_yaw);
+    auto yaw_sin = sinf(initial_yaw);
+    auto dx = speed.linear.x * yaw_cos;
+    auto dy = speed.linear.x * yaw_sin;
+    if (is_holo) {
+      dx -= speed.linear.y * yaw_sin;
+      dy += speed.linear.y * yaw_cos;
+    }
+    pose.pose.position.x += dx*dt;
+    pose.pose.position.y += dy*dt;
+    initial_yaw += speed.angular.z * pred_dt;
+    tf2::Quaternion quat;
+    quat.setRPY(0.0, 0.0, initial_yaw);
+    pose.pose.orientation = quat;
+  }
+  /**
    * @brief Whether the motion model is holonomic, using Y axis
    * @return Bool If holonomic
    */

--- a/nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp
+++ b/nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp
@@ -28,6 +28,8 @@
 #include "nav2_mppi_controller/models/constraints.hpp"
 
 #include "nav2_mppi_controller/tools/parameters_handler.hpp"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+#include "tf2/utils.hpp"
 
 namespace mppi
 {
@@ -169,27 +171,29 @@ public:
   /**
    * @brief With input pose, speed find the vehicle's output pose at pred_dt in future
    * @param pose pose of the robot
-   * @param speed speed 
+   * @param speed speed
    * @param pred_dt time in future to predict the pose
   */
-  virtual void predictFuture(geometry_msgs::msg::PoseStamped& pose, geometry_msgs::msg::Twist& speed, float pred_dt)
+  virtual void predictFuture(
+    geometry_msgs::msg::PoseStamped & pose,
+    geometry_msgs::msg::Twist & speed, float pred_dt)
   {
     const bool is_holo = isHolonomic();
     auto initial_yaw = static_cast<float>(tf2::getYaw(pose.pose.orientation));
     auto yaw_cos = cosf(initial_yaw);
     auto yaw_sin = sinf(initial_yaw);
-    auto dx = speed.linear.x * yaw_cos;
-    auto dy = speed.linear.x * yaw_sin;
+    auto dx = static_cast<float>(speed.linear.x) * yaw_cos;
+    auto dy = static_cast<float>(speed.linear.x) * yaw_sin;
     if (is_holo) {
       dx -= speed.linear.y * yaw_sin;
       dy += speed.linear.y * yaw_cos;
     }
-    pose.pose.position.x += dx*dt;
-    pose.pose.position.y += dy*dt;
+    pose.pose.position.x += dx * pred_dt;
+    pose.pose.position.y += dy * pred_dt;
     initial_yaw += speed.angular.z * pred_dt;
     tf2::Quaternion quat;
     quat.setRPY(0.0, 0.0, initial_yaw);
-    pose.pose.orientation = quat;
+    pose.pose.orientation = tf2::toMsg(quat);
   }
   /**
    * @brief Whether the motion model is holonomic, using Y axis

--- a/nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp
+++ b/nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp
@@ -28,9 +28,8 @@
 #include "nav2_mppi_controller/models/constraints.hpp"
 
 #include "nav2_mppi_controller/tools/parameters_handler.hpp"
-#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+#include "nav2_util/geometry_utils.hpp"
 #include "tf2/utils.hpp"
-
 namespace mppi
 {
 
@@ -191,9 +190,7 @@ public:
     pose.pose.position.x += dx * pred_dt;
     pose.pose.position.y += dy * pred_dt;
     initial_yaw += speed.angular.z * pred_dt;
-    tf2::Quaternion quat;
-    quat.setRPY(0.0, 0.0, initial_yaw);
-    pose.pose.orientation = tf2::toMsg(quat);
+    pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(initial_yaw);
   }
   /**
    * @brief Whether the motion model is holonomic, using Y axis

--- a/nav2_mppi_controller/src/optimizer.cpp
+++ b/nav2_mppi_controller/src/optimizer.cpp
@@ -311,24 +311,28 @@ void Optimizer::prepare(
     // Clamp to physically achievable range so prediction never exceeds dynamics.
     const auto & c = settings_.constraints;
     const double dt = settings_.controller_period;
+    float max_delta_vx = dt * c.ax_max;
+    float min_delta_vx = dt * c.ax_min;
+    float max_delta_vy = dt * c.ay_max;
+    float min_delta_vy = dt * c.ay_min;
+    float max_delta_wz = dt * c.az_max;
     state_.speed = robot_speed;
-    state_.speed.linear.x = std::clamp(
-      last_command_vel_.linear.x,
-      robot_speed.linear.x + dt * c.ax_min,
-      robot_speed.linear.x + dt * c.ax_max);
-    state_.speed.angular.z = std::clamp(
-      last_command_vel_.angular.z,
-      robot_speed.angular.z - dt * c.az_max,
-      robot_speed.angular.z + dt * c.az_max);
+    auto robot_last_speed = state_.speed;
+    state_.speed.linear.x = utils::clampVelocityByAccel(
+      robot_speed.linear.x, last_command_vel_.linear.x,  min_delta_vx, max_delta_vx);
+    state_.speed.angular.z = utils::clampVelocityByAccel(
+      robot_speed.angular.z, last_command_vel_.angular.z,  -max_delta_wz, max_delta_wz);
     if (isHolonomic()) {
-      state_.speed.linear.y = std::clamp(
-        last_command_vel_.linear.y,
-        robot_speed.linear.y + dt * c.ay_min,
-        robot_speed.linear.y + dt * c.ay_max);
+      state_.speed.linear.y = utils::clampVelocityByAccel(
+      robot_speed.linear.y, last_command_vel_.linear.y,  min_delta_vy, max_delta_vy);
     }
   }
 
   state_.pose = robot_pose;
+  // Predict the robot pose at 1*dt in future
+  // assuming robot is executing current observed speed until dt in future,
+  motion_model_->predictFuture(state_.pose,robot_last_speed,dt);
+
   state_.local_path_length = nav2_util::geometry_utils::calculate_path_length(plan);
   path_ = utils::toTensor(plan);
   costs_.setZero(settings_.batch_size);

--- a/nav2_mppi_controller/src/optimizer.cpp
+++ b/nav2_mppi_controller/src/optimizer.cpp
@@ -303,9 +303,9 @@ void Optimizer::prepare(
   const geometry_msgs::msg::Pose & goal,
   nav2_core::GoalChecker * goal_checker)
 {
+  state_.pose = robot_pose;
   if (settings_.open_loop) {
     state_.speed = last_command_vel_;
-    state_.pose = robot_pose;
   } else {
     // Predict state one controller_period forward toward the last command to compensate
     // for the latency between measurement and when this command will take effect.
@@ -314,8 +314,6 @@ void Optimizer::prepare(
     const double dt = settings_.controller_period;
     float max_delta_vx = dt * c.ax_max;
     float min_delta_vx = dt * c.ax_min;
-    float max_delta_vy = dt * c.ay_max;
-    float min_delta_vy = dt * c.ay_min;
     float max_delta_wz = dt * c.az_max;
     state_.speed = robot_speed;
     auto robot_last_speed = state_.speed;
@@ -324,12 +322,12 @@ void Optimizer::prepare(
     state_.speed.angular.z = utils::clampVelocityByAccel(
       robot_speed.angular.z, last_command_vel_.angular.z, -max_delta_wz, max_delta_wz);
     if (isHolonomic()) {
+      float max_delta_vy = dt * c.ay_max;
+      float min_delta_vy = dt * c.ay_min;
       state_.speed.linear.y = utils::clampVelocityByAccel(
       robot_speed.linear.y, last_command_vel_.linear.y, min_delta_vy, max_delta_vy);
     }
-    state_.pose = robot_pose;
     // Predict the robot pose at 1*dt in future
-    // assuming robot is executing current observed speed until dt in future,
     motion_model_->predictFuture(state_.pose, robot_last_speed, dt);
   }
 

--- a/nav2_mppi_controller/src/optimizer.cpp
+++ b/nav2_mppi_controller/src/optimizer.cpp
@@ -305,6 +305,7 @@ void Optimizer::prepare(
 {
   if (settings_.open_loop) {
     state_.speed = last_command_vel_;
+    state_.pose = robot_pose;
   } else {
     // Predict state one controller_period forward toward the last command to compensate
     // for the latency between measurement and when this command will take effect.
@@ -319,19 +320,19 @@ void Optimizer::prepare(
     state_.speed = robot_speed;
     auto robot_last_speed = state_.speed;
     state_.speed.linear.x = utils::clampVelocityByAccel(
-      robot_speed.linear.x, last_command_vel_.linear.x,  min_delta_vx, max_delta_vx);
+      robot_speed.linear.x, last_command_vel_.linear.x, min_delta_vx, max_delta_vx);
     state_.speed.angular.z = utils::clampVelocityByAccel(
-      robot_speed.angular.z, last_command_vel_.angular.z,  -max_delta_wz, max_delta_wz);
+      robot_speed.angular.z, last_command_vel_.angular.z, -max_delta_wz, max_delta_wz);
     if (isHolonomic()) {
       state_.speed.linear.y = utils::clampVelocityByAccel(
-      robot_speed.linear.y, last_command_vel_.linear.y,  min_delta_vy, max_delta_vy);
+      robot_speed.linear.y, last_command_vel_.linear.y, min_delta_vy, max_delta_vy);
     }
+    state_.pose = robot_pose;
+    // Predict the robot pose at 1*dt in future
+    // assuming robot is executing current observed speed until dt in future,
+    motion_model_->predictFuture(state_.pose, robot_last_speed, dt);
   }
 
-  state_.pose = robot_pose;
-  // Predict the robot pose at 1*dt in future
-  // assuming robot is executing current observed speed until dt in future,
-  motion_model_->predictFuture(state_.pose,robot_last_speed,dt);
 
   state_.local_path_length = nav2_util::geometry_utils::calculate_path_length(plan);
   path_ = utils::toTensor(plan);

